macros/src/trans.rs

```diff
@@ -0,0 +1,486 @@
+    Ident::new("rtfm")
+}
+
+pub fn app(app: &App, ceilings: &Ceilings) -> Tokens {
+    let mut main = vec![];
+    let mut root = vec![];
+
+    super::trans::init(app, &mut main, &mut root);
+    super::trans::idle(app, ceilings, &mut main, &mut root);
+    super::trans::resources(app, ceilings, &mut root);
+    super::trans::tasks(app, ceilings, &mut root);
+
+    root.push(quote! {
+        fn main() {
+            #(#main)*
+        }
+    });
+
+    quote!(#(#root)*)
+}
+
+fn init(app: &App, main: &mut Vec<Tokens>, root: &mut Vec<Tokens>) {
+    let device = &app.device;
+    let krate = krate();
+
+    let mut fields = vec![];
+    let mut exprs = vec![];
+    let mut lifetime = None;
+    for name in &app.init.resources {
+        lifetime = Some(quote!('a));
+
+        if let Some(resource) = app.resources.get(name) {
+            let ty = &resource.ty;
+
+            fields.push(quote! {
+                pub #name: &'a mut #ty,
+            });
+
+            exprs.push(quote! {
+                #name: &mut *super::#name.get(),
+            });
+        } else {
+            fields.push(quote! {
+                pub #name: &'a mut ::#device::#name,
+            });
+
+            exprs.push(quote! {
+                #name: &mut *::#device::#name.get(),
+            });
+        }
+    }
+
+    root.push(quote! {
+        mod init {
+            #[allow(non_snake_case)]
+            pub struct Resources<#lifetime> {
+                #(#fields)*
+            }
+
+            impl<#lifetime> Resources<#lifetime> {
+                pub unsafe fn new() -> Self {
+                    Resources {
+                        #(#exprs)*
+                    }
+                }
+            }
+        }
+    });
+
+    let mut exceptions = vec![];
+    let mut interrupts = vec![];
+    for (name, task) in &app.tasks {
+        match task.kind {
+            Kind::Exception => {
+                if exceptions.is_empty() {
+                    exceptions.push(quote! {
+                        let scb = #device::SCB.borrow(cs);
+                    });
+                }
+
+                let priority = task.priority;
+                exceptions.push(quote! {
+                    let prio_bits = #device::NVIC_PRIO_BITS;
+                    let hw = ((1 << prio_bits) - #priority) << (8 - prio_bits);
+                    scb.shpr[rtfm::Exception::#name.nr() - 4].write(hw);
+                });
+            }
+            Kind::Interrupt { enabled } => {
+                if interrupts.is_empty() {
+                    interrupts.push(quote! {
+                        let nvic = #device::NVIC.borrow(cs);
+                    });
+                }
+
+                let priority = task.priority;
+                interrupts.push(quote! {
+                    let prio_bits = #device::NVIC_PRIO_BITS;
+                    let hw = ((1 << prio_bits) - #priority) << (8 - prio_bits);
+                    nvic.set_priority(#device::Interrupt::#name, hw);
```

Comment:
you mean adding a `set_logical_priority` to NVIC or adding a `logical2hw` function to map a logical priority to a "hardware" value that can be directly written to an NVIC register